File: full-cfr/com/htmake/reader/db/JSONTable.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  io.vertx.core.json.JsonArray
 *  io.vertx.core.json.JsonObject
 *  kotlin.Metadata
 *  kotlin.Unit
 *  kotlin.jvm.functions.Function1
 *  kotlin.jvm.functions.Function2
 *  kotlin.jvm.functions.Function3
 *  kotlin.jvm.internal.Intrinsics
 *  org.jetbrains.annotations.NotNull
 *  org.jetbrains.annotations.Nullable
 */
package com.htmake.reader.db;

import com.htmake.reader.db.DB;
import com.htmake.reader.utils.ExtKt;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import java.util.ArrayList;
import java.util.List;
import kotlin.Metadata;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import kotlin.jvm.functions.Function3;
import kotlin.jvm.internal.Intrinsics;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u0000V\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010\u000b\n\u0002\b\u0002\n\u0002\u0010\u0000\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u0011\n\u0002\b\u0002\u0018\u0000*\u0004\b\u0000\u0010\u00012\b\u0012\u0004\u0012\u0002H\u00010\u0002B\u0015\u0012\u0006\u0010\u0003\u001a\u00020\u0004\u0012\u0006\u0010\u0005\u001a\u00020\u0004\u00a2\u0006\u0002\u0010\u0006J\u001c\u0010\u0007\u001a\u00020\b2\u0012\u0010\t\u001a\u000e\u0012\u0004\u0012\u00020\u000b\u0012\u0004\u0012\u00020\f0\nH\u0016J7\u0010\r\u001a\u0004\u0018\u00018\u0000\"\b\b\u0001\u0010\u000e*\u00020\u000f2\u0006\u0010\u0010\u001a\u00020\u00042\u0006\u0010\u0011\u001a\u0002H\u000e2\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00028\u00000\u0013H\u0016\u00a2\u0006\u0002\u0010\u0014J\b\u0010\u0015\u001a\u00020\u0016H\u0016J\b\u0010\u0017\u001a\u00020\bH\u0016JQ\u0010\u0017\u001a\u00020\b2\u0006\u0010\u0011\u001a\u00028\u00002 \u0010\u0018\u001a\u001c\u0012\u0004\u0012\u00028\u0000\u0012\u0004\u0012\u00020\f\u0012\u0004\u0012\u00020\u0016\u0012\u0004\u0012\u00020\b\u0018\u00010\u00192\u0018\u0010\t\u001a\u0014\u0012\u0004\u0012\u00020\u000b\u0012\u0004\u0012\u00028\u0000\u0012\u0004\u0012\u00020\f0\u001aH\u0016\u00a2\u0006\u0002\u0010\u001bJW\u0010\u001c\u001a\u00020\b2\f\u0010\u0011\u001a\b\u0012\u0004\u0012\u00028\u00000\u001d2 \u0010\u0018\u001a\u001c\u0012\u0004\u0012\u00028\u0000\u0012\u0004\u0012\u00020\f\u0012\u0004\u0012\u00020\u0016\u0012\u0004\u0012\u00020\b\u0018\u00010\u00192\u0018\u0010\t\u001a\u0014\u0012\u0004\u0012\u00020\u000b\u0012\u0004\u0012\u00028\u0000\u0012\u0004\u0012\u00020\f0\u001aH\u0016\u00a2\u0006\u0002\u0010\u001e\u00a8\u0006\u001f"}, d2={"Lcom/htmake/reader/db/JSONTable;", "T", "Lcom/htmake/reader/db/DB;", "userNameSpace", "", "name", "(Ljava/lang/String;Ljava/lang/String;)V", "delete", "", "checker", "Lkotlin/Function1;", "Lio/vertx/core/json/JsonObject;", "", "findBy", "P", "", "field", "value", "clazz", "Ljava/lang/Class;", "(Ljava/lang/String;Ljava/lang/Object;Ljava/lang/Class;)Ljava/lang/Object;", "readAll", "Lio/vertx/core/json/JsonArray;", "save", "onCheckEnd", "Lkotlin/Function3;", "Lkotlin/Function2;", "(Ljava/lang/Object;Lkotlin/jvm/functions/Function3;Lkotlin/jvm/functions/Function2;)V", "saveMulti", "", "([Ljava/lang/Object;Lkotlin/jvm/functions/Function3;Lkotlin/jvm/functions/Function2;)V", "reader-pro"})
public final class JSONTable<T>
extends DB<T> {
    public JSONTable(@NotNull String userNameSpace, @NotNull String name) {
        Intrinsics.checkNotNullParameter((Object)userNameSpace, (String)"userNameSpace");
        Intrinsics.checkNotNullParameter((Object)name, (String)"name");
        super(userNameSpace, name);
    }

    @Override
    @NotNull
    public JsonArray readAll() {
        String[] stringArray = new String[]{"data", this.getUserNameSpace(), this.getName()};
        JsonArray dataList = ExtKt.asJsonArray(ExtKt.getStorage$default(stringArray, null, 2, null));
        if (dataList == null) {
            dataList = new JsonArray();
        }
        this.setCachedValue(dataList);
        return dataList;
    }

    @Override
    @Nullable
    public <P> T findBy(@NotNull String field, @NotNull P value, @NotNull Class<T> clazz) {
        Intrinsics.checkNotNullParameter((Object)field, (String)"field");
        Intrinsics.checkNotNullParameter(value, (String)"value");
        Intrinsics.checkNotNullParameter(clazz, (String)"clazz");
        JsonArray dataList = this.readAll();
        int n = 0;
        int n2 = dataList.size();
        if (n < n2) {
            do {
                int i;
                Object objValue;
                if (!value.equals(objValue = dataList.getJsonObject(i = n++).getValue(field))) continue;
                return (T)dataList.getJsonObject(i).mapTo(clazz);
            } while (n < n2);
        }
        return null;
    }

    @Override
    public void save(T value, @Nullable Function3<? super T, ? super Boolean, ? super JsonArray, Unit> onCheckEnd, @NotNull Function2<? super JsonObject, ? super T, Boolean> checker) {
        Intrinsics.checkNotNullParameter(checker, (String)"checker");
        JsonArray dataList = this.readAll();
        int existIndex = -1;
        int n = 0;
        int n2 = dataList.size();
        if (n < n2) {
            do {
                int i = n++;
                JsonObject jsonObject = dataList.getJsonObject(i);
                Intrinsics.checkNotNullExpressionValue((Object)jsonObject, (String)"dataList.getJsonObject(i)");
                if (!((Boolean)checker.invoke((Object)jsonObject, value)).booleanValue()) continue;
                existIndex = i;
                break;
            } while (n < n2);
        }
        if (onCheckEnd != null) {
            onCheckEnd.invoke(value, (Object)(existIndex >= 0 ? 1 : 0), (Object)dataList);
        }
        if (existIndex >= 0) {
            List list2 = dataList.getList();
            list2.set(existIndex, JsonObject.mapFrom(value));
            dataList = new JsonArray(list2);
        } else {
            dataList.add(JsonObject.mapFrom(value));
        }
        this.setCachedValue(dataList);
        this.save();
    }

    @Override
    public void saveMulti(@NotNull T[] value, @Nullable Function3<? super T, ? super Boolean, ? super JsonArray, Unit> onCheckEnd, @NotNull Function2<? super JsonObject, ? super T, Boolean> checker) {
        Intrinsics.checkNotNullParameter(value, (String)"value");
        Intrinsics.checkNotNullParameter(checker, (String)"checker");
        JsonArray dataList = this.readAll();
        int existIndex = -1;
        T[] TArray = value;
        int n = 0;
        int n2 = TArray.length;
        while (n < n2) {
            T j = TArray[n];
            ++n;
            int n3 = 0;
            int n4 = dataList.size();
            if (n3 < n4) {
                do {
                    int i = n3++;
                    JsonObject jsonObject = dataList.getJsonObject(i);
                    Intrinsics.checkNotNullExpressionValue((Object)jsonObject, (String)"dataList.getJsonObject(i)");
                    if (!((Boolean)checker.invoke((Object)jsonObject, j)).booleanValue()) continue;
                    existIndex = i;
                    break;
                } while (n3 < n4);
            }
            if (onCheckEnd != null) {
                onCheckEnd.invoke(j, (Object)(existIndex >= 0 ? 1 : 0), (Object)dataList);
            }
            if (existIndex >= 0) {
                dataList.set(existIndex, JsonObject.mapFrom(j));
                continue;
            }
            dataList.add(JsonObject.mapFrom(j));
        }
        this.setCachedValue(dataList);
        this.save();
    }

    @Override
    public void delete(@NotNull Function1<? super JsonObject, Boolean> checker) {
        Intrinsics.checkNotNullParameter(checker, (String)"checker");
        JsonArray dataList = this.readAll();
        int n = 0;
        List removeIndexList = new ArrayList();
        n = 0;
        int n2 = dataList.size();
        if (n < n2) {
            do {
                int i = n++;
                JsonObject jsonObject = dataList.getJsonObject(i);
                Intrinsics.checkNotNullExpressionValue((Object)jsonObject, (String)"dataList.getJsonObject(i)");
                if (!((Boolean)checker.invoke((Object)jsonObject)).booleanValue()) continue;
                removeIndexList.add(i);
            } while (n < n2);
        }
        if (removeIndexList.size() > 0) {
            JsonArray newList = new JsonArray();
            n2 = 0;
            int n3 = dataList.size();
            if (n2 < n3) {
                do {
                    int i;
                    if (removeIndexList.contains(i = n2++)) continue;
                    newList.add(dataList.getJsonObject(i));
                } while (n2 < n3);
            }
            dataList = newList;
        }
        this.setCachedValue(dataList);
        this.save();
    }

    @Override
    public void save() {
        String[] stringArray = new String[]{"data", this.getUserNameSpace(), this.getName()};
        ExtKt.saveStorage$default(stringArray, this.getCachedValue(), false, null, 12, null);
    }
}

File: full-vineflower/com/htmake/reader/db/JSONTable.kt
package com.htmake.reader.db

import com.htmake.reader.utils.ExtKt
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import java.util.ArrayList

public class JSONTable<T>(userNameSpace: String, name: String) : DB(userNameSpace, name) {
   public override fun readAll(): JsonArray {
      var dataList: JsonArray = ExtKt.asJsonArray(
         ExtKt.getStorage$default(new java.lang.String[]{"data", this.getUserNameSpace(), this.getName()}, null, 2, null)
      );
      if (dataList == null) {
         dataList = new JsonArray();
      }

      this.setCachedValue(dataList);
      return dataList;
   }

   public override fun <P : Any> findBy(field: String, value: P, clazz: Class<Any>): Any? {
      val dataList: JsonArray = this.readAll();
      var var5: Int = 0;
      val var6: Int = dataList.size();
      if (0 < var6) {
         do {
            val i: Int = var5++;
            if (value.equals(dataList.getJsonObject(i).getValue(field))) {
               return dataList.getJsonObject(i).mapTo(clazz);
            }
         } while (var5 < var6);
      }

      return null;
   }

   public override fun save(value: Any, onCheckEnd: ((Any, Boolean, JsonArray) -> Unit)?, checker: (JsonObject, Any) -> Boolean) {
      var dataList: JsonArray = this.readAll();
      var existIndex: Int = -1;
      var list: Int = 0;
      val var7: Int = dataList.size();
      if (0 < var7) {
         do {
            val i: Int = list++;
            val var9: JsonObject = dataList.getJsonObject(i);
            if (checker.invoke(var9, value) as java.lang.Boolean) {
               existIndex = i;
               break;
            }
         } while (list < var7);
      }

      if (onCheckEnd != null) {
         onCheckEnd.invoke(value, existIndex >= 0, dataList);
      }

      if (existIndex >= 0) {
         val var10: java.util.List = dataList.getList();
         var10.set(existIndex, JsonObject.mapFrom(value));
         dataList = new JsonArray(var10);
      } else {
         dataList.add(JsonObject.mapFrom(value));
      }

      this.setCachedValue(dataList);
      this.save();
   }

   public override fun saveMulti(value: Array<Any>, onCheckEnd: ((Any, Boolean, JsonArray) -> Unit)?, checker: (JsonObject, Any) -> Boolean) {
      val dataList: JsonArray = this.readAll();
      var existIndex: Int = -1;
      val var6: Array<Any> = value;
      var var7: Int = 0;
      val var8: Int = value.length;

      while (var7 < var8) {
         val j: Any = var6[var7];
         var7++;
         var var10: Int = 0;
         val var11: Int = dataList.size();
         if (0 < var11) {
            do {
               val i: Int = var10++;
               val var13: JsonObject = dataList.getJsonObject(i);
               if (checker.invoke(var13, j) as java.lang.Boolean) {
                  existIndex = i;
                  break;
               }
            } while (var10 < var11);
         }

         if (onCheckEnd != null) {
            onCheckEnd.invoke(j, existIndex >= 0, dataList);
         }

         if (existIndex >= 0) {
            dataList.set(existIndex, JsonObject.mapFrom(j));
         } else {
            dataList.add(JsonObject.mapFrom(j));
         }
      }

      this.setCachedValue(dataList);
      this.save();
   }

   public override fun delete(checker: (JsonObject) -> Boolean) {
      var dataList: JsonArray = this.readAll();
      val removeIndexList: java.util.List = new ArrayList();
      var var8: Int = 0;
      var var5: Int = dataList.size();
      if (0 < var5) {
         do {
            val i: Int = var8++;
            val ix: JsonObject = dataList.getJsonObject(i);
            if (checker.invoke(ix) as java.lang.Boolean) {
               removeIndexList.add(i);
            }
         } while (var8 < var5);
      }

      if (removeIndexList.size() > 0) {
         val var9: JsonArray = new JsonArray();
         var5 = 0;
         val var11: Int = dataList.size();
         if (0 < var11) {
            do {
               val var12: Int = var5++;
               if (!removeIndexList.contains(var12)) {
                  var9.add(dataList.getJsonObject(var12));
               }
            } while (var5 < var11);
         }

         dataList = var9;
      }

      this.setCachedValue(dataList);
      this.save();
   }

   public override fun save() {
      ExtKt.saveStorage$default(new java.lang.String[]{"data", this.getUserNameSpace(), this.getName()}, this.getCachedValue(), false, null, 12, null);
   }
}
